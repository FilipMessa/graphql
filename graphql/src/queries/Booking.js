// @flow

import { GraphQLID, GraphQLNonNull } from 'graphql';
import GraphQLBooking from '../outputs/Booking';

import type { GraphqlContextType } from '../services/GraphqlContext';
import type { BookingsItem } from '../types/Booking';

export default {
  type: GraphQLBooking, // may be null (doesn't exist)
  description: 'Flight booking by ID.',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
  },
  resolve: async (
    ancestor: mixed,
    { id }: Object,
    { dataLoader }: GraphqlContextType,
  ): Promise<BookingsItem> => dataLoader.bookings.loadItem(id),
};
