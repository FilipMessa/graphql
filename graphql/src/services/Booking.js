import jsonMock from '../../tests/bookings.json';

// const url = `https://booking-api.skypicker.com/api/v0.1/users/self/bookings?token=${token}`;

export default () => jsonMock; // FIXME: replace with API call
