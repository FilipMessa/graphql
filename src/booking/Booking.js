// @flow

import type { DepartureArrival, Leg } from '../flight/Flight';
import type { Price } from '../common/types/Price';
import type { AllowedBaggage } from './Baggage';
import type { TripData } from './types/outputs/Trip';

/**
 * This is fetched from 'bookings' endpoint.
 */
export type BookingType =
  | 'BookingOneWay'
  | 'BookingReturn'
  | 'BookingMulticity';

export type BookingsItem = {
  id: number,
  arrival: DepartureArrival,
  departure: DepartureArrival,
  legs: Array<Leg>,
  price: Price,
  created: Date,
  authToken: string,
  status: string,
  type: BookingType,
  passengerCount: number,
  trips?: TripData[],
  inbound?: TripData,
  outbound?: TripData,
  passengers: Passenger[],
};

export type TravelDocument = {|
  idNumber: string,
  expiration: number,
|};

export type Visa = {|
  info: string,
  status: string,
  timestamp: number,
  country: string,
|};

// This type could be expanded with health and passport information
export type TravelInfo = {|
  visa: Visa[],
|};

export type Pkpass = {|
  +flightNumber: string,
  +url: string,
  +passenger?: Passenger,
|};

export type Passenger = {|
  id: number,
  firstname: string,
  lastname: string,
  insuranceType: string,
  title: string,
  birthday: string,
  nationality: string,
  travelDocument: TravelDocument,
  travelInfo: TravelInfo[],
  pkpasses: $ReadOnlyArray<Pkpass> | null,
|};

export type BookedService = {|
  category: string,
  status: string,
|};

export type ContactDetails = {|
  phone: string,
  email: string,
  passenger: ?Passenger,
|};

/**
 * This is additionally fetched from 'booking/12..' endpoint if needed.
 */
export type Booking = BookingsItem & {
  allowedBaggage: AllowedBaggage,
  assets: BookingAssets,
  bookedServices: BookedService[],
  contactDetails: ContactDetails,
  allowedToChangeFlights: ?number,
  onlineCheckinIsAvailable: boolean,
  insurancePrices: InsurancePrice[],
};

export type BoardingPass = {|
  +boardingPassUrl: mixed,
  +flightNumber: string,
  +availableAt: ?string,
  +leg: ?Leg,
  +pkpasses: $ReadOnlyArray<Pkpass>,
|};

export type BookingAssets = {
  ticketUrl: ?string,
  invoiceUrl: ?string,
  boardingPasses: BoardingPass[],
  legs: Leg[],
};

export type InsurancePrice = {|
  +type: 'travel_basic' | 'travel_plus' | 'none',
  +price: Price | null,
|};
