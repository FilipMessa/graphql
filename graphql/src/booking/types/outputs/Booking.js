// @flow

import { GraphQLInt, GraphQLList, GraphQLObjectType } from 'graphql';
import { globalIdField } from '../../../common/services/OpaqueIdentifier';

import GraphQLRouteStop from '../../../flight/types/outputs/RouteStop';
import GraphQLAllowedBaggage from './AllowedBaggage';
import GraphQLLeg from '../../../flight/types/outputs/Leg';
import GraphQLBookingAssets from './BookingAssets';
import GraphQLBookingStatus from '../enums/BookingStatus';

import type { AllowedBaggage } from '../../Baggage';
import type { Booking, BookingAssets } from '../../Booking';
import type { DepartureArrival, Leg } from '../../../flight/Flight';
import type { GraphqlContextType } from '../../../common/services/GraphqlContext';

export default new GraphQLObjectType({
  name: 'Booking',
  fields: {
    id: globalIdField(),

    databaseId: {
      type: GraphQLInt,
      description: 'Internal database ID.',
      deprecationReason: 'Use id field instead.',
      resolve: ({ id }: Booking): number => id,
    },

    allowedBaggage: {
      type: GraphQLAllowedBaggage,
      resolve: async (
        { id }: Booking,
        params: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<AllowedBaggage> => {
        const { allowedBaggage } = await dataLoader.booking.load(id);
        return allowedBaggage;
      },
    },

    arrival: {
      type: GraphQLRouteStop,
      resolve: ({ arrival }: Booking): DepartureArrival => arrival,
    },

    assets: {
      type: GraphQLBookingAssets,
      description: 'Static assets related to this booking.',
      resolve: async (
        { id }: Booking,
        params: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<BookingAssets> => {
        const { assets } = await dataLoader.booking.load(id);
        return assets;
      },
    },

    departure: {
      type: GraphQLRouteStop,
      resolve: ({ departure }: Booking): DepartureArrival => departure,
    },

    legs: {
      type: new GraphQLList(GraphQLLeg),
      description: 'Flight segments, e.g. stopover, change of aircraft, etc.',
      resolve: ({ legs }: Booking): Leg[] => legs,
    },

    status: {
      type: GraphQLBookingStatus,
      resolve: ({ status }: Booking): string => status,
    },
  },
});
