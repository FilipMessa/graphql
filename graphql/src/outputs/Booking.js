// @flow

import { GraphQLID, GraphQLInt, GraphQLList, GraphQLObjectType } from 'graphql';

import GraphQLRouteStop from './RouteStop';
import GraphQLAllowedBaggage from './AllowedBaggage';
import GraphQLLeg from './Leg';
import GraphQLBookingAssets from './BookingAssets';
import { toGlobalId } from '../services/OpaqueIdentifier';

import type { AllowedBaggage } from '../types/Baggage';
import type { Booking, BookingAssets } from '../types/Booking';
import type { DepartureArrival, Leg } from '../types/Flight';
import type { GraphqlContextType } from '../services/GraphqlContext';

export default new GraphQLObjectType({
  name: 'Booking',
  fields: {
    id: {
      type: GraphQLID,
      resolve: ({ id }: Booking): string => toGlobalId('booking', id),
    },

    databaseId: {
      type: GraphQLInt,
      description: 'Internal database ID.',
      deprecationReason: 'Use id field instead.',
      resolve: ({ id }: Booking): number => id,
    },

    allowedBaggage: {
      type: GraphQLAllowedBaggage,
      resolve: async (
        { id }: Booking,
        params: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<AllowedBaggage> => {
        const { allowedBaggage } = await dataLoader.booking.load(id);
        return allowedBaggage;
      },
    },

    arrival: {
      type: GraphQLRouteStop,
      resolve: ({ arrival }: Booking): DepartureArrival => arrival,
    },

    assets: {
      type: GraphQLBookingAssets,
      description: 'Static assets related to this booking.',
      resolve: async (
        { id }: Booking,
        params: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<BookingAssets> => {
        const { assets } = await dataLoader.booking.load(id);
        return assets;
      },
    },

    departure: {
      type: GraphQLRouteStop,
      resolve: ({ departure }: Booking): DepartureArrival => departure,
    },

    legs: {
      type: new GraphQLList(GraphQLLeg),
      description: 'Flight segments, e.g. stopover, change of aircraft, etc.',
      resolve: ({ legs }: Booking): Leg[] => legs,
    },
  },
});
