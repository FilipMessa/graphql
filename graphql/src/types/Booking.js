// @flow

import { GraphQLID, GraphQLInt, GraphQLList, GraphQLObjectType } from 'graphql';

import GraphQLRouteStop from './RouteStop';
import GraphQLAllowedBaggage from './AllowedBaggage';
import GraphQLLeg from './Leg';
import { toGlobalId } from '../services/OpaqueIdentifier';

import type {
  DepartureArrivalType,
  AllowedBaggageType,
  BookingType,
  LegType,
} from '../Entities';
import type { GraphqlContextType } from '../services/GraphqlContext';

export default new GraphQLObjectType({
  name: 'Booking',
  fields: {
    id: {
      type: GraphQLID,
      resolve: ({ id }: BookingType): string => toGlobalId('booking', id),
    },

    databaseId: {
      type: GraphQLInt,
      resolve: ({ id }: BookingType): number => id,
    },

    allowedBaggage: {
      type: GraphQLAllowedBaggage,
      resolve: async (
        { id }: BookingType,
        params: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<AllowedBaggageType> => {
        const { allowedBaggage } = await dataLoader.booking.load(id);
        return allowedBaggage;
      },
    },

    arrival: {
      type: GraphQLRouteStop,
      resolve: ({ arrival }: BookingType): DepartureArrivalType => arrival,
    },

    departure: {
      type: GraphQLRouteStop,
      resolve: ({ departure }: BookingType): DepartureArrivalType => departure,
    },

    legs: {
      type: new GraphQLList(GraphQLLeg),
      resolve: ({ legs }: BookingType): Array<LegType> => legs,
    },
  },
});
