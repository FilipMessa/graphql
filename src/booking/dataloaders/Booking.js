// @flow

import DataLoader from 'dataloader';
import { get } from '../../common/services/HttpRequest';
import Config from '../../../config/application';
import { sanitizeDetail } from './ApiSanitizer';

import type { Booking } from '../Booking';
import type BookingsLoader from './Bookings';

export default function createInstance(
  accessToken: ?string,
  bookingsLoader: BookingsLoader,
) {
  return new DataLoader(
    (ids: $ReadOnlyArray<number | string>) =>
      batchLoad(accessToken, bookingsLoader)(ids),
    {
      cacheKeyFn: key => parseInt(key, 10),
    },
  );
}

function batchLoad(
  accessToken: ?string,
  bookingsLoader: BookingsLoader,
): ($ReadOnlyArray<number | string>) => Promise<Array<Object>> {
  if (typeof accessToken !== 'string') {
    return () => Promise.reject(new Error('Undefined access token'));
  }
  return ids => Promise.all(ids.map(id => fetch(parseInt(id), bookingsLoader)));
}

async function fetch(bid, bookingsLoader): Promise<Booking> {
  const { authToken } = await bookingsLoader.loadItem(bid);
  const data = await get(Config.restApiEndpoint.singleBooking(bid, authToken));

  return sanitizeDetail(data);
}
