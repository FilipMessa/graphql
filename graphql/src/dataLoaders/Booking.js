// @flow

import DataLoader from 'dataloader';
import request from '../services/HttpRequest';
import Config from '../../config/application';
import { sanitizeDetail } from '../queries/booking/ApiSanitizer';
import type { BookingType } from '../Entities';
import type BookingsLoader from '../dataLoaders/Bookings';

export default function createInstance(
  accessToken: ?string,
  bookingsLoader: BookingsLoader,
) {
  return new DataLoader(
    (ids: Array<number | string>) =>
      batchLoad(accessToken, bookingsLoader)(ids),
    {
      cacheKeyFn: key => parseInt(key, 10),
    },
  );
}

function batchLoad(
  accessToken: ?string,
  bookingsLoader: BookingsLoader,
): (Array<number | string>) => Promise<Array<Object>> {
  if (typeof accessToken !== 'string') {
    return () => Promise.reject(new Error('Undefined access token'));
  }
  return ids => Promise.all(ids.map(id => fetch(parseInt(id), bookingsLoader)));
}

async function fetch(bid, bookingsLoader): Promise<BookingType> {
  const { authToken } = await bookingsLoader.loadItem(bid);
  const data = await request(
    Config.restApiEndpoint.singleBooking(bid, authToken),
  );

  return sanitizeDetail(data);
}
