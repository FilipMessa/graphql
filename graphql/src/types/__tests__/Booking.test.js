import Booking from '../Booking';

describe('id field', () => {
  const id = Booking.getFields().id;
  it('should be non-null ID type', () => {
    expect(id.type.toString()).toBe('ID!');
  });
});

describe('databaseId field', () => {
  const databaseId = Booking.getFields().databaseId;
  it('should be non-null Int type', () => {
    expect(databaseId.type.toString()).toBe('Int!');
  });
});

describe('arrival field', () => {
  const arrival = Booking.getFields().arrival;
  it('should be non-null Arrival type', () => {
    expect(arrival.type.toString()).toBe('Arrival!');
  });
});

describe('departure field', () => {
  const departure = Booking.getFields().departure;
  it('should be non-null Departure type', () => {
    expect(departure.type.toString()).toBe('Departure!');
  });
});

describe('flights field', () => {
  const flights = Booking.getFields().flights;
  it('should be non-null list of non-null Flight types', () => {
    expect(flights.type.toString()).toBe('[Flight!]!');
  });
});
