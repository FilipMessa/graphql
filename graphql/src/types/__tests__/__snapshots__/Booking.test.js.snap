// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking type should have valid fields 1`] = `
Object {
  "allowedBaggage": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "allowedBaggage",
    "resolve": [Function],
    "type": "AllowedBaggage!",
  },
  "arrival": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "arrival",
    "resolve": [Function],
    "type": "RouteStop!",
  },
  "databaseId": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "databaseId",
    "resolve": [Function],
    "type": "Int!",
  },
  "departure": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "departure",
    "resolve": [Function],
    "type": "RouteStop!",
  },
  "id": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "id",
    "resolve": [Function],
    "type": "ID!",
  },
  "legs": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "legs",
    "resolve": [Function],
    "type": "[Leg!]!",
  },
}
`;
